lib/Conversion/AffineToPipeline/AffineToPipeline.cpp

```diff
@@ -57,20 +57,93 @@ struct AffineToPipeline : public AffineToPipelineBase<AffineToPipeline> {
   LogicalResult populateOperatorTypes(SmallVectorImpl<AffineForOp> &loopNest);
   LogicalResult solveSchedulingProblem(SmallVectorImpl<AffineForOp> &loopNest);
   LogicalResult createPipelinePipeline(SmallVectorImpl<AffineForOp> &loopNest);
+  LogicalResult constantCSE();
+  LogicalResult unrollAndForwardStores();
 
   CyclicSchedulingAnalysis *schedulingAnalysis;
 };
 
+// This forwards stores to loads for a very specific, particular pattern: store-load pairs on memrefs that accumulate
+// accumulates results in the body of a for loop:
+//  affine.for %arg3 = 0 to 8 {
+//    %3 = affine.load %arg0[%arg1, %arg3] : memref<1x8xi32>
+//    %4 = affine.load %1[%arg3, %arg2] : memref<8x8xi32>
+//    %5 = affine.load %2[%arg1, %arg2] : memref<1x8xi32>
+//    %6 = arith.muli %3, %4 : i32
+//    %7 = arith.addi %5, %6 : i32
+//    affine.store %7, %2[%arg1, %arg2] : memref<1x8xi32>
+//  }
+// After unrolling this looks like
+//  ...
+//  %281 = arith.muli %278, %279 : i32
+//  %282 = arith.addi %280, %281 : i32
+//  affine.store %282, %1[%arg1, %241] : memref<1x8xi32>
+//  %283 = affine.apply #map6(%c0_0)
+//  %284 = affine.load %arg0[%arg1, %283] : memref<1x8xi32>
+//  %285 = affine.load %0[%283, %241] : memref<8x8xi32>
+//  %286 = affine.load %1[%arg1, %241] : memref<1x8xi32>
+//  %287 = arith.muli %284, %285 : i32
+// where we observe that affine.store %282 can be forwarded to %284 = affine.load.
+// Thus, in this method we search for such pairs - a load and its immediately prior store.
+static LogicalResult forwardFullyUnrolledStoreToLoad(AffineReadOpInterface loadOp,
+                                                     SmallVectorImpl<Operation *> &opsToErase) {
+  Operation *prevOp = loadOp->getPrevNode();
+  while (prevOp) {
+    auto storeOp = dyn_cast<AffineWriteOpInterface>(prevOp);
+    if (!storeOp) {
+      prevOp = prevOp->getPrevNode();
+      continue;
+    }
+
+    MemRefAccess srcAccess(storeOp);
+    MemRefAccess dstAccess(loadOp);
+
+    if (srcAccess != dstAccess)
+      return failure();
+
+    // Since we forward only from the immediately prior store, we verify that there are neither intervening
+    // stores nor intervening loads in between.
+    if (!mlir::hasNoInterveningEffect<MemoryEffects::Write>(storeOp, loadOp)
+        || !mlir::hasNoInterveningEffect<MemoryEffects::Read>(storeOp, loadOp))
```

Comment:
Since you decided to use hasNoInterveningEffect, which checks things recursively, using only MemoryEffects::Write should be enough.